commons/serialization/src/main/scala/sbt/pickling/json/JSONPickleFormat.scala

```diff
@@ -50,33 +53,123 @@ package json {
       }
     }
   }
+  object JSONPickleFormat {
+    private[json] val TYPE_TAG_FIELD = "$type"
+    private[json] val DYNAMIC_KEY_FIELD = "$keys"
+    private[json] val REF_ID_FIELD = "$ref"
+
+    private[json] def isSpecialField(name: String): Boolean =
+      (TYPE_TAG_FIELD == name) || (DYNAMIC_KEY_FIELD == name) || (REF_ID_FIELD == name)
+    private[json] def isElidedField(name: String): Boolean =
+      (DYNAMIC_KEY_FIELD == name)
+  }
 
-  class JSONPickleBuilder(format: JSONPickleFormat, buf: Output[String]) extends PBuilder with PickleTools {
-    private var nindent = 0
-    private def indent() = nindent += 1
-    private def unindent() = nindent -= 1
-    private var pendingIndent = false
-    private var lastIsBrace = false
-    private var lastIsBracket = false
-    private def append(s: String) = {
-      val sindent = if (pendingIndent) "  " * nindent else ""
-      buf.put(sindent + s)
-      pendingIndent = false
-      val trimmed = s.trim
-      if (trimmed.nonEmpty) {
-        val lastChar = trimmed.last
-        lastIsBrace = lastChar == '{'
-        lastIsBracket = lastChar == '['
+  sealed trait BuilderState {
+    def previous: BuilderState
+  }
+  case class CollectionState(val previous: BuilderState, numElements: Int, hasInput: Boolean) extends BuilderState
+  case class RawEntryState(previous: BuilderState, picklee: Any, hints: Hints, var wasCollectionOrMap: Boolean = false) extends BuilderState
+  case class MapEntryState(val previous: BuilderState, picklee: Any, hints: Hints) extends BuilderState
+  case class RefEntryState(val previous: BuilderState) extends BuilderState
+  object EmptyState extends BuilderState {
+    def previous = this
+  }
+
+  // A slow implementation of of a pickle builder
+  // This uses a TON of branch statements to ensure the builder is in the correct state for any call
+  // and to programatically enforce constraints of SPickler implementations.
+  // We use this just to verify our own picklers.
+  class VerifyingJSONPickleBuilder(format: JSONPickleFormat, buf: Output[String]) extends PBuilder with PickleTools {
+    import JSONPickleFormat._
+    var state: BuilderState = EmptyState
+    //(tag.key startsWith "scala.Option[")
+    private def isJValue(tag: FastTypeTag[_]): Boolean =
+      (tag.key startsWith "org.json4s.JsonAST.")
+    // Here we get notified of object/value-like things.
+    override def beginEntry(picklee: Any): PBuilder = withHints { hints =>
+      // Here we check to see if we need to serialize a reference.  These are used to avoid circular object
+      // dependencies for picklers which have circluarly-references objects.
+      if (hints.oid != -1) {
+        buf.put("{\"" + REF_ID_FIELD + "\":" + hints.oid + "}")
```

Comment:
Agree on interpolation, although buf.put vs append feels a style preference.  We don't indent because the wire protocol doesn't need indent.  We can use json4s to pretty print later.